plugin/passwords.go

```diff
@@ -0,0 +1,51 @@
+package plugin
+
+import (
+	"context"
+	"strings"
+
+	"github.com/hashicorp/vault/sdk/helper/base62"
+)
+
+var (
+	// Per https://en.wikipedia.org/wiki/Password_strength#Guidelines_for_strong_passwords
+	minimumLengthOfComplexString = 8
+
+	passwordComplexityPrefix = "?@09AZ"
+	pwdFieldTmpl             = "{{PASSWORD}}"
+)
+
+type passwordGenerator interface {
+	GeneratePasswordFromPolicy(ctx context.Context, policyName string) (password string, err error)
+}
+
+// GeneratePassword from the password configuration. This will either generate based on a password policy
+// or from the provided formatter. The formatter/length options are deprecated.
+func GeneratePassword(ctx context.Context, passConf passwordConf, generator passwordGenerator) (password string, err error) {
+	err = passConf.validate()
+	if err != nil {
+		return "", err
+	}
+
+	if passConf.PolicyName != "" {
+		return generator.GeneratePasswordFromPolicy(ctx, passConf.PolicyName)
+	}
+	return generateDeprecatedPassword(passConf.Formatter, passConf.Length)
+}
+
+func generateDeprecatedPassword(formatter string, totalLength int) (string, error) {
+	pwd, err := base62.Random(totalLength)
+	if err != nil {
+		return "", err
+	}
+	if formatter == "" {
+		pwd = passwordComplexityPrefix + pwd
+		return pwd[:totalLength], nil
+	}
+	return strings.Replace(formatter, pwdFieldTmpl, pwd[:lengthOfPassword(formatter, totalLength)], 1), nil
```

Comment:
Should `lengthOfPassword(formatter, totalLength)` be checked before trying to uses it as an index?

Reply:
This is existing code (moved around). I'm pretty sure this is safe because `pwd` has a length of `totalLength` and `lengthOfPassword(formatter, totalLength)` returns a length that's strictly less than or equal to `totalLength`.

Reply:
Thinking about it more, I think I can make the code more readable and remove this possibility entirely.